dbms/src/Functions/geohashCoverBox.cpp

```diff
@@ -0,0 +1,166 @@
+        ColumnString::Offsets & res_strings_offsets = res_strings.getOffsets();
+
+        UInt32 pos = 0;
+        for (size_t row = 0; row < total_rows; ++row)
+        {
+            const Float64 lon_min_value = lon_min->getElement(row);
+            const Float64 lat_min_value = lat_min->getElement(row);
+            const Float64 lon_max_value = lon_max->getElement(row);
+            const Float64 lat_max_value = lat_max->getElement(row);
+            const PrecisionType precision_value = GeoUtils::geohashPrecision(precision->getElement(row));
+
+            const auto positions = GeoUtils::geohashCoverBox(lon_min_value, lat_min_value,
+                                                          lon_max_value, lat_max_value,
+                                                          precision_value, max_elements);
+
+            const auto array_size = positions.size();
+
+            res_strings_offsets.reserve(res_strings_offsets.size() + array_size);
+            res_strings_chars.resize(res_strings_chars.size() + array_size * (precision_value + 1));
+
+            char * out = reinterpret_cast<char *>(res_strings_chars.data());
+
+            for (size_t i = 0; i < array_size; ++i)
+            {
+                const Float64 longitude_value = positions[i].first;
+                const Float64 latitude_value = positions[i].second;
+
+                const size_t encoded_size = GeoUtils::geohashEncode(longitude_value, latitude_value, precision_value, &out[pos]);
+
+                pos += encoded_size;
+                out[pos] = '\0';
+                ++pos;
+                res_strings_offsets.push_back(pos);
+            }
+            res_offsets.push_back(array_size);
+
+            if (!res_strings_offsets.empty() && res_strings_offsets.back() != res_strings_chars.size())
+            {
+                throw Exception("String column size mismatch (internal logical error)", ErrorCodes::LOGICAL_ERROR);
+            }
+        }
+        if (!res_offsets.empty() && res_offsets.back() != res_strings.size())
+            throw Exception("Arrary column size mismatch (internal logical error)", ErrorCodes::LOGICAL_ERROR);
+
+        result = std::move(col_res);
+        return true;
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        const IColumn * lon_min = block.getByPosition(arguments[0]).column.get();
+        const IColumn * lat_min = block.getByPosition(arguments[1]).column.get();
+        const IColumn * lon_max = block.getByPosition(arguments[2]).column.get();
+        const IColumn * lat_max = block.getByPosition(arguments[3]).column.get();
+        const IColumn * prec =    block.getByPosition(arguments[4]).column.get();
+        ColumnPtr & res = block.getByPosition(result).column;
+
+        if (tryExecute<Float32, Float32, Float32, Float32, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res) ||
```

Comment:
If these four template parameters are always going to be the same, one is enough.
Also we can just determine the type from the first argument and call the right function. We've already checked that types are the same in getReturnTypeImpl.

Reply:
well, I don't know how convenient having all arguments of the same type to the user, so I've left these template arguments as a backup plan. On the other hand I don't want to blow up code size with N specializations of that template.

Reply:
Done